src/coreclr/nativeaot/Runtime/threadstore.inl

```diff
@@ -1,7 +1,12 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-EXTERN_C DECLSPEC_THREAD ThreadBuffer tls_CurrentThread;
+#ifdef _MSC_VER
+// a workaround to prevent tls_CurrentThread from becoming dynamically checked/initialized.
+__declspec(selectany) __declspec(thread) ThreadBuffer tls_CurrentThread;
+#else
+EXTERN_C __thread ThreadBuffer tls_CurrentThread;
```

Comment:
Does non-Windows pay the penalty for TLS init checks? (ie similar workaround is not possible on non-Windows)

Reply:
From the discussion of the original fix (https://github.com/dotnet/runtime/pull/33347), I gathered that even on windows the workaround was not needed until some version of MSVC. 
It is probably more a function of the c runtime (libc, musl, etc). It should be fairly easy to check what happens on libc. 
I also wonder if there are differences on arm64.

Reply:
Linux/glibc does not require a similar fix. There are no checks when accessing TLS.

Here is the disassembly for reference.

Linux/glibc   x64
```asm
System.Collections.Tests`::RhGetCurrentManagedThreadId():
    0x5555554d5db0 <+0>:  pushq  %rbp
    0x5555554d5db1 <+1>:  movq   %rsp, %rbp
    0x5555554d5db4 <+4>:  movq   %fs:0x0, %rax
    0x5555554d5dbd <+13>: leaq   -0x100(%rax), %rax
    0x5555554d5dc4 <+20>: movl   0xc0(%rax), %eax
    0x5555554d5dca <+26>: popq   %rbp
    0x5555554d5dcb <+27>: retq
```

Linux/glibc arm64
```asm
System.Collections.Tests`::RhGetCurrentManagedThreadId():
    0xaaaaaab87400 <+0>:  stp    x29, x30, [sp, #-0x10]!
    0xaaaaaab87404 <+4>:  mov    x29, sp
    0xaaaaaab87408 <+8>:  adrp   x0, 4173
    0xaaaaaab8740c <+12>: ldr    x0, [x0, #0xa30]
    0xaaaaaab87410 <+16>: nop
    0xaaaaaab87414 <+20>: nop
    0xaaaaaab87418 <+24>: mrs    x8, TPIDR_EL0
    0xaaaaaab8741c <+28>: add    x8, x8, x0
    0xaaaaaab87420 <+32>: ldr    w0, [x8, #0xc0]
    0xaaaaaab87424 <+36>: ldp    x29, x30, [sp], #0x10
    0xaaaaaab87428 <+40>: ret
```

MSVC    (with the fix)
```asm
    return ThreadStore::RawGetCurrentThread()->GetManagedThreadId();
00007FF6A99D7AE0  mov         ecx,dword ptr [_tls_index (07FF6A9CB3A28h)]  
00007FF6A99D7AE6  mov         rax,qword ptr gs:[58h]  
00007FF6A99D7AEF  mov         edx,10h  
00007FF6A99D7AF4  mov         rax,qword ptr [rax+rcx*8]  
00007FF6A99D7AF8  mov         eax,dword ptr [rax+rdx+0C0h]  
}
00007FF6A99D7AFF  ret  
```